libdnf5/rpm/solv/goal_private.cpp

```diff
@@ -262,61 +268,55 @@ bool GoalPrivate::limit_installonly_packages(libdnf5::solv::IdQueue & job, Id ru
             // and negative for conflicts (conflicts with another package or dependency
             // conflicts = dependencies cannot be met).
             if (libsolv_solver.get_decisionlevel(p) > 0) {
-                q.push_back(p);
+                marked_for_install_providers_map.add_unsafe(p);
             } else {
-                Solvable * solvable = spool.id2solvable(p);
-                if (!spool.is_installed(solvable)) {
-                    available_unused_providers.push_back(solvable);
-                }
+                available_unused_providers_map.add_unsafe(p);
             }
         }
-        if (q.size() <= static_cast<int>(installonly_limit)) {
-            continue;
-        }
-        for (int k = 0; k < q.size(); ++k) {
-            Id id = q[k];
-            Solvable * s = spool.id2solvable(id);
-            if (spool->installed != s->repo) {
-                installing.push_back(id);
-                break;
-            }
+    }
+
+    libdnf5::solv::IdQueue marked_for_install_providers;
+    for (const auto & pkg_id : marked_for_install_providers_map) {
+        marked_for_install_providers.push_back(pkg_id);
+    }
+
+    std::vector<Solvable *> available_unused_providers;
+    for (const auto & pkg_id : available_unused_providers_map) {
+        Solvable * solvable = spool.id2solvable(pkg_id);
+        if (!spool.is_installed(solvable)) {
+            available_unused_providers.push_back(solvable);
         }
-        if (!installing.size()) {
+    }
+
+    const InstallonlyCmpData installonly_cmp_data{spool, running_kernel};
+    marked_for_install_providers.sort(&installonly_cmp, &installonly_cmp_data);
+    std::sort(available_unused_providers.begin(), available_unused_providers.end(), name_solvable_cmp_key);
+
+    libdnf5::solv::IdQueue same_names;
+    while (!marked_for_install_providers.empty()) {
```

Comment:
It would be great to add a comment what while loop does.

Reply:
Added.